File: 3112.cpp
// 3112. Minimum Time to Visit Disappearing Nodes
// There is an undirected graph of n nodes. You are given a 2D array edges, where edges[i] = [ui, vi, lengthi] describes an edge between node ui and node vi with a traversal time of lengthi units.
// Additionally, you are given an array disappear, where disappear[i] denotes the time when the node i disappears from the graph and you won't be able to visit it.
// Note that the graph might be disconnected and might contain multiple edges.
// Return the array answer, with answer[i] denoting the minimum units of time required to reach node i from node 0. If node i is unreachable from node 0 then answer[i] is -1.

class Solution {
public:
    struct Edge{

        int u;
        int v;
        int l;

        Edge(int u, int v, int l): u(u), v(v), l(l){}
    };

    vector<int> minimumTime(int n, vector<vector<int>>& edges, vector<int>& disappear) {
        vector<int> ans(n, -1);
        vector<int> dist(n, INT_MAX);
        dist[0] = 0;

        vector<vector<Edge>> adj(n);
        for(const auto& edge: edges)
        {
            int u = edge[0];
            int v = edge[1];
            int l = edge[2];
            adj[u].emplace_back(u, v, l);
            adj[v].emplace_back(v, u, l);
        }

        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        pq.emplace(0, 0);

        while(!pq.empty())
        {
            auto [cur_dist, u] = pq.top();
            pq.pop();
            if((cur_dist > dist[u] || cur_dist >= disappear[u]))
            {
                continue;
            }
                for(const Edge& edge : adj[u])
                {
                    int v = edge.v;
                    int d = cur_dist + edge.l;
                    if(d < dist[v] && d < disappear[v])
                    {
                        dist[v] = d;
                        pq.emplace(d, v);
                    }
                }
            
        }

         for (int i = 0; i < n; ++i) 
        {
            if (dist[i] < disappear[i]) 
            {
            ans[i] = dist[i];
            }
        }
        ans[0] = 0; 
        return ans;

    }
};

File: 3112.py
from typing import List
from collections import defaultdict
import heapq
import math

class Solution:
    def minimumTime(self, n: int, edges: List[List[int]], disappear: List[int]) -> List[int]:
        graph = defaultdict(list)
        for u, v, length in edges:
            graph[u].append((v, length))
            graph[v].append((u, length))
       
        visited = [False] * n
        min_times = [math.inf] * n
        min_times[0] = 0
        priority_queue = [(0, 0)]

        while priority_queue != []:
            current_time, current_node = heapq.heappop(priority_queue)

            if visited[current_node] == True:
                continue

            visited[current_node] = True

            for neighbor, edge_length in graph[current_node]:
                if visited[neighbor] == True:
                    continue

                new_time = current_time + edge_length
                if new_time <= disappear[neighbor] and new_time < min_times[neighbor]:
                    min_times[neighbor] = new_time
                    heapq.heappush(priority_queue, (new_time, neighbor))

        result = []
        for time in min_times:
            result.append(time)

        return result

